content/posts/best-practices-for-cloudbees-core-jenkins-on-kubernetes/securely-using-cloud-services-from-jenkins-kubernetes-agents.md

```diff
@@ -0,0 +1,271 @@
+
+The default configuration for the [Jenkins Kubernetes Cloud plugin](https://github.com/jenkinsci/kubernetes-plugin) uses the same Namespace and Kubernetes Service Account as the Jenkins Master it is configured for - when the Jenkins Master is also running on Kubernetes. This is typically the case with Core on Kubernetes and also for OSS Jenkins. To fully leverage binding IAM permissions to Kubernetes Service Accounts we must revisit how we set-up and use the Jenkins Kubernetes Clouds for agents.
+
+#### Create a unique Kubernetes Namespace and ServiceAccount for each Managed Master and Kubernetes Cloud
+Placing each of your Jenkins Masters into their own Kubernetes Namespaces provides and extra layer of security that isn't just limited to binding IAM credentials - it also protects Kubernetes Secrets from other Jenkins Masters and this will provide a more secure integration for managing Jenkins credentials with JCasC and Kubernetes Secrets. Creating a unique Kubernetes Cloud per Jenkins Master can be [managed more easily with JCasC](https://github.com/kypseli/workshop-mm-jcasc/blob/ops/jcasc.yml), as we will manage the Kubernetes cloud configuration and the Jenkins credential used to connect to Kubernetes for the Master specific cloud. The Jenkins credential used for the Kubernetes Cloud configuration will be the Kubernetes Service Account token stored in Jenkins as a Secret Text credential - stored at the Master level (not CloudBees Core Operations Center), so only that Jenkins Master has access to it. The Kubernetes Service Account token should be managed as a Kubernetes Secret in the same Namespace as the Managed Master is created so it can be dynamically injected into the JCasC configuration for that Master. No other Core Managed Masters will have access to this Kubernetes Secret.
+
+>NOTE: This approach depends on managing what Kuberentes Namespaces are used by Managed Masters. You must not allow untrusted users to have access to configure Managed Masters on your Core Operations Center, otherwise they could use a Namespace they shouldn't have access to - that is why [I prefer to do it as code](https://github.com/kypseli/demo-mm-jcasc/blob/cloud-run/groovy/createManagedMaster.groovy).
+
+#### Example: Using Workload Identity with a Core Managed Master to Deploy a Container Image to Google Cloud Run
+
+1. [Enable Workload Identity for your GKE cluster](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#enable_workload_identity_on_an_existing_cluster) - this only has to be done once per cluster.
```

Comment:
In addition to enabling Workload Identity on the cluster itself, I also had to [enable it on my existing node pool](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#migrate_workloads_to_workload_identity) - might be worth calling out here.